x/lightclient/ante/ibc_msg_update_client.go

```diff
@@ -0,0 +1,93 @@
+	if canonicalClient != msg.ClientId {
+		return nil // The client is not a rollapp's canonical client. Continue with default behaviour.
+	}
+	clientMessage, err := ibcclienttypes.UnpackClientMessage(msg.ClientMessage)
+	if err != nil {
+		return nil
+	}
+	header, ok := clientMessage.(*ibctm.Header)
+	if !ok {
+		return nil
+	}
+	// Check if there are existing block descriptors for the given height of client state
+	height := header.TrustedHeight
+	stateInfo, err := i.rollappKeeper.FindStateInfoByHeight(ctx, rollappID, height.GetRevisionHeight())
+	if err != nil {
+		// No BDs found for given height.
+		// Will accept the update optimistically
+		// But also save the blockProposer address with the height for future verification
+		blockProposer := header.Header.ProposerAddress
+		i.lightClientKeeper.SetConsensusStateSigner(ctx, msg.ClientId, height.GetRevisionHeight(), blockProposer)
+		return nil
+	}
+	bd, _ := stateInfo.GetBlockDescriptor(height.GetRevisionHeight())
+
+	ibcState := types.IBCState{
+		Root:               header.Header.GetAppHash(),
+		ValidatorsHash:     header.Header.ValidatorsHash,
+		NextValidatorsHash: header.Header.NextValidatorsHash,
+		Timestamp:          header.Header.Time,
+	}
+	sequencerPubKey, err := i.lightClientKeeper.GetSequencerPubKey(ctx, stateInfo.Sequencer)
+	if err != nil {
+		return err
+	}
+	rollappState := types.RollappState{
+		BlockSequencer:  sequencerPubKey,
+		BlockDescriptor: bd,
+	}
+	// Check that BD for next block exists in same stateinfo
+	if stateInfo.ContainsHeight(height.GetRevisionHeight() + 1) {
+		rollappState.NextBlockSequencer = sequencerPubKey
+		rollappState.NextBlockDescriptor, _ = stateInfo.GetBlockDescriptor(height.GetRevisionHeight() + 1)
+	} else {
+		// next BD does not exist in this state info, check the next state info
+		nextStateInfo, found := i.rollappKeeper.GetStateInfo(ctx, rollappID, stateInfo.GetIndex().Index+1)
+		if found {
+			nextSequencerPk, err := i.lightClientKeeper.GetSeqeuncerHash(ctx, nextStateInfo.Sequencer)
+			if err != nil {
+				return err
+			}
+			rollappState.NextBlockSequencer = nextSequencerPk
+			rollappState.NextBlockDescriptor = nextStateInfo.GetBDs().BD[0]
+		} else {
+			// if next state info does not exist, then we can't verify the next block valhash.
+			// Will accept the update optimistically
+			// But also save the blockProposer address with the height for future verification
+			// When the corresponding state info is submitted by the sequencer, will perform the verification
+			blockProposer := header.Header.ProposerAddress
+			i.lightClientKeeper.SetConsensusStateSigner(ctx, msg.ClientId, height.GetRevisionHeight(), blockProposer)
+			return nil
+		}
+	}
+	// Ensure that the ibc header is compatible with the existing rollapp state
+	// If it's not, we error and prevent the MsgUpdateClient from being processed
+	err = types.CheckCompatibility(ibcState, rollappState)
+	if err != nil {
+		return err
+	}
```

Comment:
can you simplify this

In pseudocode

```
MsgUpdateClient(h) {
 bdH, okH := rollappKeeper.FindStateInfoByHeight(h).Bd(h)
 bdHPlus1, okHPlus1 := rollappKeeper.FindStateInfoByHeight(h+1).Bd(h+1)
 if !(okH&&okHPlus1) {
     // accept optimistic
 }else{
 checkCompatability(..)
 }
}
```

We don't care about micro optimization in reducing binary search calls in some cases

Reply:
also the db will optimize